metaflow/package/mfenv.py

```diff
@@ -0,0 +1,867 @@
+        handling_cls = cls._get_mfenv_class(mfenv_info)
+        if handling_cls:
+            return handling_cls._get_archive_filename(
+                archive, name, mfenv_info, is_meta, packaging_backend
+            )
+        # Backward compatible way of accessing all special files. Prior to MFEnv, they
+        # were stored at the TL of the archive. There is no distinction between code and
+        # config.
+        real_name = meta_file_name(name)
+        if packaging_backend.cls_has_member(archive, real_name):
+            return real_name
+        return None
+
+    @classmethod
+    def get_archive_content(
+        cls,
+        archive: tarfile.TarFile,
+        name: Union[MetaFile, str],
+        is_meta: Optional[bool] = None,
+        packaging_backend=TarPackagingBackend,
+    ) -> Optional[bytes]:
+        """
+        Get the content of a file in archive. The content is returned as
+        bytes
+
+        Parameters
+        ----------
+        name : Union[MetaFile, str]
+            Filename to look for. If it is a MetaFile, it is assumed to be relative
+            to the configuration directory. If it is a string, it is assumed to be
+            relative to the code directory.
+        is_meta : bool, optional, default None
+            If None, the default behavior above is assumed. If True, the file will be
+            searched for in the configuration directory and if False, it will be
+            searched for in the code directory.
+
+        Returns
+        -------
+        Optional[bytes]
+            The binary content of the file -- None if there is no such file.
+        """
+        file_to_extract = cls.get_archive_filename(
+            archive, name, is_meta, packaging_backend
+        )
+        if file_to_extract:
+            return packaging_backend.cls_extract_member(archive, file_to_extract)
+        return None
+
+    def __init_subclass__(cls, version_id, **kwargs) -> None:
+        super().__init_subclass__(**kwargs)
+        if version_id in MFEnv._mappings:
+            raise ValueError(
+                "Version ID %s already exists in MFEnv mappings "
+                "-- this is a bug in Metaflow." % str(version_id)
+            )
+        MFEnv._mappings[version_id] = cls
+
+    @classmethod
+    def _get_mfenv_class(cls, info: Optional[Dict[str, Any]]):
+        if info is None:
+            return None
+        if "version" not in info:
+            raise MetaflowException(
+                "Invalid package -- missing version in info: %s" % info
+            )
+        version = info["version"]
+        if version not in cls._mappings:
+            raise MetaflowException(
+                "Invalid package -- unknown version %s in info: %s" % (version, info)
+            )
+
+        return cls._mappings[version]
+
+    @classmethod
+    def _extract_archive_mfenv_info(
+        cls, archive: Any, packaging_backend=TarPackagingBackend
```

Comment:
any methods that need access to archive can be safely moved under MetaflowPackage?